data_structures/binary_search_tree.cpp

```diff
@@ -1,172 +1,296 @@
 /**
  * \file
  * \brief A simple tree implementation using structured nodes
- *
- * \todo update code to use C++ STL library features and OO structure
- * \warning This program is a poor implementation - C style - and does not
- * utilize any of the C++ STL features.
  */
+#include <algorithm>  // test code
+#include <cassert>
 #include <iostream>
+#include <memory>
+#include <queue>
+#include <vector>
 
-struct node {
-    int val;
-    node *left;
-    node *right;
-};
+/**
+ * @namespace data_structure
+ * @brief Implementations of various data structures
+ */
+namespace data_structure {
+/**
+ * @namespace data_structure::BST
+ * @brief Implementation of Binary Search Tree data structure
+ */
+namespace BST {
 
-struct Queue {
-    node *t[100];
-    int front;
-    int rear;
+/** An struct node type.
+ *  Define a element for tree insert/remove operation
+ */
+struct node {
+    int val{};                   /**< int value of a node struct*/
+    std::shared_ptr<node> left;  /**< left subtree pointer */
+    std::shared_ptr<node> right; /**< right subtree pointer */
+    //~node() { std::cout << "node: destructor, val = " << val << '\n'; }
 };
 
-Queue queue;
-
-void enqueue(node *n) { queue.t[queue.rear++] = n; }
-
-node *dequeue() { return (queue.t[queue.front++]); }
+/** insert a node to tree
+ * with greater value to right subtree, smaller value node to left subtree
+ * \param[in] root root node of a tree
+ * \param[in] x a node with value to be insert
+ */
+void Insert(node* root, int x) {
+    if (root == NULL) {
+        std::cout << " \n invalid root node = " << root << std::endl;
+        std::cout << " \n please call Insert() with valid root node!!! \n";
+        return;
+    }
 
-void Insert(node *n, int x) {
-    if (x < n->val) {
-        if (n->left == NULL) {
-            node *temp = new node;
-            temp->val = x;
-            temp->left = NULL;
-            temp->right = NULL;
-            n->left = temp;
+    if (x < root->val) {
```

Comment:
what if we have a new value which is equal to the current root node? you have not handled this case when the given value is equal to the value of the current root node. we can have duplicate values in BST as far as i know there is no such boundation to not to have duplicate elements. so consider this case too.

Reply:
wouldn't that condition be treated by line# 61?

Reply:
For a test, I executed this program with this input: 
1. root = 5, insert = 5, insert = 5
2. display breadth first: `5 5 5`
3. insert = 2
4. display: `5 2 5 5`
5. insert = 7
6. display breadth first: `5 2 5 5 7`

Is this the expectation?

Reply:
absolutely! thanks for pointing, my bad luck i don't know why i forgot that there is an else block

Reply:
> For a test, I executed this program with the following steps:
> 
> 1. root = 5, insert = 5, insert = 5
> 2. display breadth first: `5 5 5`
> 3. insert = 2
> 4. display: `5 2 5 5`
> 5. insert = 7
> 6. display breadth first: `5 2 5 5 7`
> 7. Post order depth first: `2 7 5 5 5`
> 8. In-order depth first: `2 5 5 5 7`
> 9. Pre-order depth first: `5 2 5 5 7`
> 
> Is this the expectation?

yes these results are good